gui/upgrade.h

```diff
@@ -19,13 +19,17 @@
 
  */
 
-#include "babeldata.h"
-#include "format.h"
-#include <QDateTime>
-#include <QUrl>
+#include <QtCore/QDateTime>                 // for QDateTime
+#include <QtCore/QList>                     // for QList
+#include <QtCore/QObject>                   // for QObject
+#include <QtCore/QString>                   // for QString
+#include <QtCore/QUrl>                      // for QUrl
+#include <QtNetwork/QNetworkAccessManager>  // for QNetworkAccessManager
+#include <QtNetwork/QNetworkReply>          // for QNetworkReply
+#include <QtWidgets/QWidget>                // for QWidget
+#include "babeldata.h"                      // for BabelData
+#include "format.h"                         // for Format
 
-class QNetworkAccessManager;
```

Comment:
Forward decls when we need only the pointer and not any member access means we can skip the #includes. We probably don't do enough of this in the porject. Is this just your script getting overambitious?

Reply:
I was swayed by https://google.github.io/styleguide/cppguide.html#Forward_Declarations
I am running include-what-you-use and specifically telling it to avoid forward declarations.

Reply:
That's compelling.
Ironic because my mentally cached version of this documented went the other way. Compile times were dominating development and Qt's headers were some of the worst. We spent a ridiculous amount of effort tracking down scoping issues that were exposed when the project essentially did blah.cc: #include *.cc just so the tokenizer and parser didn't have to chomp on 0000's of copies of <QString> for every build. Qt headers were always at the top of our build profiles.

That document has changed a few times over the years as tooling and techniques have evolved.

We traded developer pain for build times. Looks like the pendulum has gone back.

I agree, though, that we should probably follow a formal guideline on this kind of thing instead of makign up our own.